### ISSUE TITLE
Tracking issue for RFC #2116: fallible collection allocation

### ISSUE THREAD

**Issue by aturon** [rust-lang/rust#48043]
This is a tracking issue for the RFC "fallible collection allocation" (rust-lang/rfcs#2116).

**Steps:**

- [ ] Implement the RFC
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Comment by snf** [comment]
Hi, this feature will require `oom` and `try_reserve` to be implemented. I'm needing `try_reserve` and I'm willing to work (probably slowly) on it if noone else does.
I see that the previous commit by @Gankro is [here](https://github.com/rust-lang/rust/pull/43890) and seems to match the RFC description.
Is there anything else I should consider before rebasing that PR?

**Comment by snf** [comment]
I can use some mentorship on what's the best way to implement [oom=panic/abort](https://github.com/Gankro/rfcs/blob/8d425589da7a24ee88fb56f2f465326d42ae2152/text/2116-alloc-me-maybe.md#oompanic-1) . Should I emit a global variable through rustc depending on this case?

I was taking inspiration from https://github.com/rust-lang/rust/pull/32900 but there might be an easier way to emit a boolean global variable.

**Comment by aturon** [comment]
cc @Gankro, can you help mentor?

**Comment by SimonSapin** [comment]
`try_reserve` methods are implemented in Nightly, but the accepted RFC also includes the ability to make allocation failure cause a panic rather than abort and that part is not implemented. Unfortunately the RFC does not discuss how this would be implemented, other than being controlled by a flag in `Cargo.toml`. It’s not clear to me how this can work.

**Comment by SimonSapin** [comment]
@alexcrichton, how does https://github.com/rust-lang/rust/pull/51041 fit with the goal of implementing some way to opt into oom=panic?

**Comment by alexcrichton** [comment]
Ok sure, we can figure it out in the implementation details? AFAIK there's no need to revert the change as it has basically no benefit *today*. It's a trivial change to revert as well, it's one line in the standard library

**Comment by 3n-mb** [comment]
[This example in playground](https://play.rust-lang.org/?gist=30774e7d80c72c145f0ca0c17d43822c&version=nightly&mode=debug) asks for `1024*1024*1024*1000` bytes on a server, via `v.try_reserve(len)?` without giving an error.

Is this RFC implementation is not at the point, at which errors in such cases will start to show up?
Or, is this a gross over-commit by underlying OS, and this state of affairs will stay like this?
Can `try_reserve` in principle reserve only realistic things?

**Comment by SimonSapin** [comment]
When an error is returned is entirely dependent of the underlying allocator. As far as I know Linux usually overcommits, Windows doesn’t.

https://play.rust-lang.org/?gist=9424ed75fc811213c1284c5cf97c156d&version=nightly allocates 120 PB but fails at 130 PB, it is exhausting 48 bits of address space.

Even on Linux though, this API can be useful for 32-bit programs when exhausting the address space is much more realistic.

**Comment by SimonSapin** [comment]
https://github.com/rust-lang/rust/issues/43596 was the pre-RFC feature request for `oom=panic`. I’ve repurposed it as the tracking issue for that feature, since it and `try_reserved` are likely to be stabilized at different times. This leaves this issue as tracking only `try_reserve`.

**Comment by snf** [comment]
https://github.com/rust-lang/rust/pull/52420 will change CollectionAllocErr to CollectionAllocErr<AllocError>

**Comment by SimonSapin** [comment]
Will, or would. I’ve comment there that such a change in public API design merits being discussed separately from a large PR.

**Comment by Ericson2314** [comment]
Sure I do not expect this one to be merged quickly, only to provide a concrete example of what I am talking about.

**Comment by Ericson2314** [comment]
Yeah definitely. Even if it weren't for the 2018 delaying considering these things, I wouldn't expect it to be merged quick. I just make the PR to prove the feasibility and make my ideas clearer to / interactive for others.

Ignoring the benefits of allocator polymorphism itself (that would be https://github.com/rust-lang/rust/issues/42774), I see 4 big benefits:

 - Self discipline: Applications can prevent themselves statically from forgetting to handled allocations failures.

 - Race freedom: allocation divergence due to `try_reserving` insufficient space are statically prevented.

 - Expressiveness: Unclear how to do `try_reserve`-like things for "deep" data structures (those containing indirecation).

- Library flexibility: Implement functionality just once with a parameteric error type, and you get fallibility polymorphism: downstream can instantiate with `!` to get today's divergence, or `AllocError` to allow them to be explicitly handled.

I thus advocating going with error polymorphism and abandoning `try_reverve` (I rather reuse `try_`  for the `Result<(), E>` copies of every method we'll need for backwards compat). (But, my PR keeps the existing meaning of `try_reserve` so as to allow a transition period where people can try both approaches.)

**Comment by SimonSapin** [comment]
Firefox needs `try_reserve` badly enough that they forked `Vec` and `HashMap` over it. It’d be nice to stabilize.

Some of the discussion above is about `oom=panic` whose tracking has moved to https://github.com/rust-lang/rust/issues/43596. This leaves two unresolved question:

* Whether `CollectionAllocErr` should gain a type parameter, with the `AllocErr` variant holding a value of that type, in preparation for future allocation traits having an associated type for errors. I *think* we can still make that change after stabilization, if the type parameter has a default. @rust-lang/libs, does that sound accurate?

* @Ericson2314, it sounds like you’re objecting to having `try_reserve` at all, because instead we should have `try_*` variations of every existing method that allocates. However `reserve` is such a method so this proposal is a subset of yours, not contradictory. Am I missing something?

**Comment by Ericson2314** [comment]
@SimonSapin Yes if things were done my way Firefox wouldn't be `try_reserv`ing or `reserv`ing at all, but just fallibly allocate the memory they need when they need it. That's much simpler and more robust.

(There's also the more minor quibble that under the simplest version of my plan, the `Self` types are different and `try_reseve` wouldn't be callable for the regular infallible collections. But that can be changed at the cost of more associated type magic.)

The libs team paused associated allocators to prioritize the 2018 edition. That's fine with me, but then I think it only fair that Firefox wait too. I promise as soon as the gates are open I'll rush to revive @glandium's and my PRs.

**Comment by SimonSapin** [comment]
Right, `try_insert` for example would be preferable to `try_reserve` + `insert`, but it doesn’t exist yet. You know as well as I do that large changes or additions are harder to get through the process, https://github.com/rust-lang/rfcs/pull/2116 was designed deliberately to be minimal.

> I think it only fair that Firefox wait too

It’s not about who gets candy first. RFC 2116 was submitted 14 months ago and accepted 8 months ago. The `try_reserve` part has been implemented 7 months ago, stabilizing it now can be a matter of flipping a switch. Allocator-generic collections however are still in the design phase with a number of open questions, they are still gonna take significant work from multiple people.

That said, there is no real urgency for `try_reserve`. Firefox using a forked `HashMap` is unfortunate in my opinion, but it works. Stabilizing `try_reserve` only seems to me like easy incremental improvement.

**Comment by Ericson2314** [comment]
Well it sounds like we are in something closer to agreement than usual. :)

> Allocator-generic collections however are still in the design phase with a number of open questions

The code has been written and if I recall correctly works (there was one ancillary issue with trying to make default impls work well, but it is not fundamental and I believe I found a workaround beyond just requiring more boilerplate). It just needs to be read and debated.

**Comment by alexcrichton** [comment]
@SimonSapin it's been awhile since I took a look at this, but is the idea that (in the long run) every fallible method on a collection returns `CollectionAllocErr<T>` and the `T` is specialized per-method to be the "most appropriate"? In the meantime, though, we're only thinking of the `try_reserve` methods which would effectively return `CollectionAllocErr<()>` and you're thinking that we can add `<T = ()>` to the enum definition later?

One possible alternative to this is to avoid defining this type at all I think? For example unlike as we may have originally though `AllocErr` wasn't stabilized with global allocators. To that end it increases the surface area of what's to be stabilized to stabilize this feature, and it also means we're not necessarily sold on having `AllocErr` yet. These methods could, perhaps, return a `bool` for whether they succeeded or not. (this sort of depends if we think it's crucial to switch on why reservation failed, allocation and/or capacity)

I think I'd personally lean towards what you're thinking @SimonSapin by having `CollectionAllocErr` with a defaulted type parameter later on, if necessary. I might also advocate, as well, having either a non-exhaustive enum or an opaque struct with boolean accessors.

**Comment by SimonSapin** [comment]
Oh, I forgot that `AllocErr` was not stable yet. I’m fine with changing `CollectionAllocErr` to an opaque type for now. (A struct with a private field, to leave us more options later.) Even zero-size, though I’d still prefer `Result<(), _>` over `bool` even if they are semantically equivalent.

At the moment I don’t really have an opinion on whether `CollectionAllocErr` should have a type parameter. I was only responding to https://github.com/rust-lang/rust/issues/48043#issuecomment-409600561 by saying we’ll still be able to do that later if we want. (With the RFC process hat on, to make sure possible objections are not ignored.)

> This sort of depends if we think it's crucial to switch on why reservation failed, allocation and/or capacity

It looks like at least some of the code paths in Firefox track the size (and alignment) of the allocation request that failed, so the current enum might be insufficient anyway.

Alright, on further thought it may not make sense to pursue this just yet, until we figure out the rest of the story for allocation traits and error types.

**Comment by Ericson2314** [comment]
@alexcrichton So in https://github.com/rust-lang/rust/pull/52420/files I did make a `CollectionAllocErr`, but that's not really the important part, the associated error type (which would be the parameter) is.

Honestly, I'm not even sure `CollectionAllocErr` is worth it. It creates two axis with the associated type: do we hide allocation errors, and do we hide size errors? I can't imagine a case where one should cause an panic/abort and the other shouldn't. I rather have just `AllocErr` as is, or have `AllocErr` have both variants (In the memory hierarchy we can imagine that at any layer some collection/allocator gets too big, or it cannot get memory from it's backing allocator).

All this another reason that I forgot that stabilizing `try_reserve` now will tie our hands later.

**Comment by snf** [comment]
I think this is a feature that we need semi-urgently. There are projects out there that require `try_reserve` and this is not something that can be quickly replaced by a custom crate.
I like the idea of `CollectionAllocErr` with a private field but I'd like to know if anyone can think of a limitation before proceeding in implementing it.

**Comment by SimonSapin** [comment]
As far as I can tell, the error type is the only remaining blocker. How about something like this?

```rust
struct ContainerAllocError {
    layout_of_failed_allocation: Option<Layout>,
}

impl ContainerAllocError {
    fn layout_of_failed_allocation(&self) -> Option<Layout> { self.layout_of_failed_allocation }
}
```

* A struct with private fields (and no public fields), so we could later add more fields or even turn it back into an enum

* No type parameter, but we can add one later (with a default) if the `Alloc` trait grows an associated type for errors and we want to propagate custom error values

* `Container` instead of `Collection` in the name, in case we want to add `Box::try_new` at some point.
  - Should it stay in `std::collections`, or move to another module? Which one?

* Can contain the `Layout` that was passed e.g. to the `alloc()` called that failed. Having this info (or at least the size) is a requirement for Firefox, but for example the allocation strategy of `HashMap` is an implementation detail that users should not have to guess at based on `HashMap::capacity` alone.
  - Name of the accessor method to be bikeshedded.

* Speaking of bikesheds, `*Err` or `*Error`? We stabilized `LayoutErr`, but the precedent for types named `FooError` is much stronger in https://doc.rust-lang.org/std/error/trait.Error.html#implementors

**Comment by SimonSapin** [comment]
Alternative definition:

```rust
pub enum ContainerAllocError {
    CapacityOverflow,
    AllocErr {
        layout: Layout,
        #[unstable(feature = "container_alloc_error_extra")] error: AllocErr,
    },
}
```

Having a public field for a zero-size type is a bit awkward, but the key idea is that it can stay unstable while the rest of the enum is stabilized (together with `try_reserve` methods that return it).

When https://github.com/rust-lang/wg-allocators/issues/23 is resolved, based on the outcome we either remove that field:

```rust
pub enum ContainerAllocError {
    CapacityOverflow,
    AllocErr { layout: Layout },
}
```

… or make its type be a new type parameter of the enum:

```rust
pub enum ContainerAllocError<E = AllocErr> {
    CapacityOverflow,
    AllocErr {
        layout: Layout,
        error: E,
    },
}

impl SomeContainer<T, A: Alloc> {
    pub fn try_reserve(n: usize) -> Result<(), ContainerAllocError<A::Err>> {…}
}

**Comment by Ericson2314** [comment]
@SimonSapin and I disagree on whether it is superior, but my fallibility-polymorphism in https://github.com/rust-lang/rust/pull/60703 is almost ready (see how https://github.com/rust-lang/rust/pull/58457 was just rebased).

**Comment by Ericson2314** [comment]
Oh, thanks @SimonSapin for mentioning my issue in the second comment I did not see that the first time sorry. For the record, #60703 goes with the `pub enum ContainerAllocError<E = AllocErr>` approach, which is more useful because `ContainerAllocError<!>` correctly ensures only `CapacityOverflow` is possible since all all allocation failures will panic/abort.

**Comment by SimonSapin** [comment]
(At a risk of re-hashing arguments already made elsewhere, my view is that when panic/abort on allocation error is desired there is no need for `try_reserve` over `reserve`. So `ContainerAllocError<!>` is not a useful construct IMO. But the proposed unstable field to force non-exhaustive matching leaves the door open either way, so this doen’t need to be decided in this thread.)

**Comment by SimonSapin** [comment]
PR for the above: https://github.com/rust-lang/rust/pull/61780

**Comment by Ericson2314** [comment]
The PR takes into account my concerns and so looks fine to me. `Layout` is a good idea.

@SimonSapin

> (At a risk of re-hashing arguments already made elsewhere,

 I think that debate should happen in this issue? After all with my plan for fallibility/allocator polymorphism doesn't really point to `try_reserve` as written in the RFC being a good idea, even if it is technically possible to support it.

> My view is that when panic/abort on allocation error is desired there is no need for `try_reserve` over `reserve`.

My argument hinges on one key observation: the code using `reserve`/`try_reserve` may *not* want to decide whether to panic/abort or not on allocation failure, but rather leave that decision to to the user of that code. 

In general, we could have arbitrary amounts of collection and and other library code that leaves that decision to only the final executable. There's no good way to write that today without exposing `try_` and panicking versions of every method---maybe the stdlib collections have to do that anyways for backwards compat, but lets not put that burden on every other library. You can just expose the try versions and rely on users panicking themselves, but that's very annoying and tedious, and users are likely to make mistakes like panicking on too much if other non-allocation errors are possible. Put shortly, we should never tell library authors to write 2 copies of every function, or anyone to `unwrap`.

With the polymorphic alloc error, and a safe [`Result<T, !>::void_unwrap`](https://docs.rs/void/1.0.2/void/trait.ResultVoidExt.html#tymethod.void_unwrap) function, library authors can write just one function, and users can do something that, while a bit tedious, is more importantly completely safe. And it lives open the door to even less fiction with e.g. a `Result<T, !>` <=> `T` coercion.

**Comment by SimonSapin** [comment]
https://github.com/rust-lang/rust/pull/61780 has merged. I believe it resolves the remaining issues. I’ve updated the API signatures in the issue description.

Is there anything else to tweak here before stabilization? CC @Manishearth for the `hashglobe` crate, @rillian for `mp4parse_fallible`

@rfcbot fcp merge

**Comment by dtolnay** [comment]
What's the deal with the possible custom error value in TryReserveError::AllocError?

https://github.com/rust-lang/rust/blob/bdfd698f37184da42254a03ed466ab1f90e6fb6c/src/liballoc/collections/mod.rs#L59-L64

Does that involve later possibly adding a type parameter with a default value to TryReserveError? If so, that would break code that constructs CapacityOverflow because default type parameters only come into play in type position not expression position.

```rust
let _ = TryReserveError::CapacityOverflow;
```

```console
error[E0282]: type annotations needed for `TryReserveError<E>`
  --> src/main.rs:22:13
   |
22 |     let _ = TryReserveError::CapacityOverflow;
   |         -   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for `E`
   |         |
   |         consider giving this pattern the explicit type `TryReserveError<E>`, where the type parameter `E` is specified
```

**Comment by SimonSapin** [comment]
> Does that involve later possibly adding a type parameter with a default value to TryReserveError?

Correct.

> If so, that would break code that constructs CapacityOverflow because default type parameters only come into play in type position not expression position.

Ugh, that’s a good point. Back to the drawing board I guess.

@rfcbot fcp cancel

**Comment by TimDiekmann** [comment]
Another concern: If we decide to forbid zero sized allocations (https://github.com/rust-lang/wg-allocators/issues/16) the allocator would probably take a `NonZeroLayout` or similar. 
In order to not block this issue even more, I propose to mark `layout` in `AllocError` as unstable, too, until the WG found a consens.

**Comment by SimonSapin** [comment]
At this point I think there are three options here:

1. Keep blocking stabilization of this feature (which has been implemented for 19 months now) for an unbounded amount of time until the Allocators WG decides whether allocators should have an associated error type https://github.com/rust-lang/wg-allocators/issues/23. Note that WG’s repository has had very little activity in the last 5 months, after a brief surge when in was created.

2. Or take a bet that standard library collections will eventually be made allocator-generic with an associated error type, and stabilize `try_reserve` with a generic `payload` field in `TryReserveError::AllocError` that defaults to `()`.

   The risk is that if the bet turns out wrong, we’ll have a useless field and type parameter that is always `()`, in a public type.

3. Or take a bet that standard library collections will not be made allocator-generic (that feature might end up on parallel collection types, possibly outside of the standard library), or that they will but with a concrete zero-size error type, and stabilize `try_reserve` and `TryReserveError` without a `payload` field.

   The risk is that if the bet turns out wrong, we’ll have `try_reserve` methods that drop the error value it receives from the allocator.

My preference is for 2 or 3, but I don’t have a strong opinion between them. @rust-lang/libs what do you think?

**Comment by TimDiekmann** [comment]
>     Note that WG’s repository has had very little activity in the last 5 months, after a brief surge when in was created.

It is true that after the repository was opened, a few months was silent. Recently, however, there was an increase in activity again and there is a WIP PR that is currently blocking the WG from pushing anything upstream. 

So I would have suggested to make the errortype non-exhaustive. The non-exhaustiveness can still be removed later.

**Comment by alexcrichton** [comment]
I'm personally pretty wary and against stabilizing something just because progress isn't happening on it. Stabilization takes real work and has a cost. Simply because an API exists doesn't mean we should stabilize it, even if it's been sitting for <insert the amount you calculate at the time here> months. I think it's pretty bad design to get things into the standard library, then point out it's sat with no feedback for months and propose stabilization.

Fallible collections are hard in Rust. There's really no way we can escape that. At this point it still seems clear that the design work that needs to be done has not been done. It also seems like no one really knows how to get the design work done. 

I agree this would be a great feature to have, and yeah this has sat around forever as unstable. That being said the length of time something has sat is not at all a justification for stabilization in my opinion.

**Comment by SimonSapin** [comment]
> I think it's pretty bad design to get things into the standard library, then point out it's sat with no feedback for months and propose stabilization.

My understanding was that “Should `try_reserve`’s error type carry an error value that generic allocators may return” was the only remaining question that needed resolving. I wrote https://github.com/rust-lang/rust/issues/48043#issuecomment-543952770 on that basis.

But it sounds like you have concerns that are larger than that?

**Comment by alexcrichton** [comment]
It sounds like your first bullet in that comment has quite a lot of exasperation flowing through it about how slow this process is, and it also sounds like you're drawing attention to emphasize your case about how everyone else should be exasperated about it as well. I understand this process has been slow, and I think it's pretty reasonable that you're exasperated.

I don't think, however, that the exasperation is a justification for stabilization one way or another. We still do not have information about how allocators in the standard library will work out (per collection). That work just simply hasn't been done. You're proposing that we bet one way or another on this API. I think betting is fine, but I want to be very clear that exasperation at the slowness of the process is not a reason to stabilize something. It's can be a reason for stabilization because the functionality is useful enough, but it's generally unproductive in my opinion to vent such exasperation while simultaneously proposing stabilization.

I do not personally have an opinion about how this API should turn out, I don't really want to wade into the design issues of custom allocators on collections personally. I again just want to point out that the exasperation isn't particularly helpful and should never be a reason for stabilization.

**Comment by gnzlbg** [comment]
Now that some more work has happened on the design of allocators, I think I would prefer to make all methods of a particular `Vec` that allocate panic on OOM (instead of aborting), by using something like a `PanicOnOOM<MyAlloc>` allocator adapter, e.g., like:`Vec<T, PanicOnOOMAlloc<MyAlloc>>`. 

That is, instead of returning a null pointer on OOM from the alloc methods, such an alloc would just panic, allowing the user of `Vec` to catch the panics. This would make not only a handful of allocating methods fallible, but all of them. And well, if you only want "fallible" allocation for some sections of the code, you can just go from a `Vec<T, MyAlloc>` to a `Vec<T, PanicOnOOM<MyAlloc>>` for that particular section and back, and then you are guaranteed that no matter how that code gets refactored (e.g. somebody adds a call to `Vec::push` somewhere instead of `Vec::try_push`, you'll get a panic).

**Comment by SimonSapin** [comment]
The RFC that proposed `try_reserve` also separately proposed an opt-in for panic instead of abort, so that unwinding could be an additional option to recover from OOM:

* RFC: https://rust-lang.github.io/rfcs/2116-alloc-me-maybe.html
*  Tracking issue: https://github.com/rust-lang/rust/issues/43596

Under that proposal it’s `handle_alloc_error` that would change behavior. `GlobalAlloc::alloc` would still be expected (or [encouraged](https://doc.rust-lang.org/std/alloc/trait.GlobalAlloc.html#errors)) to return `null` on errors. It also does not requires [allocator-generic containers](https://github.com/rust-lang/wg-allocators/).

**Comment by malbarbo** [comment]
@SimonSapin any update?

**Comment by Ericson2314** [comment]
We now have a renewed effort at https://github.com/rust-lang/rust/pull/72314 so we can experiment with `try_*` methods across the board.

**Comment by SimonSapin** [comment]
@malbarbo I’m not sure what you’re expecting, this is the thread where further discussion about `try_reserve` should happen. As you can read above, the current status is that stabilization is blocked on deciding what extra data if any `TryReserveError::AllocErr` should carry, which ties into whether generic allocators should have an associated error types. And more generally there is not good consensus to stabilize before generic allocators overall are more settled.

@Ericson2314 https://github.com/rust-lang/rust/pull/72314 is "Stability annotations on generic parameters", which would help making containers generic on the allocator type, but how is it related to `try_*` methods? For example `Vec` can have `try_reserve` without an extra type parameter.

**Comment by TimDiekmann** [comment]
> ties into whether generic allocators should have an associated error types.

I don't think this is going to be fully resolved until we get some allocators on collections upstream and this is currently blocked by #72314.

**Comment by kornelski** [comment]
#77118 has been merged. Can stabilization of this proceed?

**Comment by TimDiekmann** [comment]
I don't really see, how #77118 is related to this? Where to you want to add an unstable generic annotation?

**Comment by kornelski** [comment]
Previous comment mentioned it was blocked on #72314, and #72314 has been replaced by #77118.

**Comment by Ericson2314** [comment]
The https://github.com/rust-lang/wg-allocators/issues/48 meta issue's "Allocator aware types" lays out the progression from #77118 to https://github.com/rust-lang/wg-allocators/issues/39, "`try_*` methods for containers".

The idea is not that `try_*` methods themselves inherently need custom allocators, but custom allocators could be leveraged to prevent the use of non-`try_*` methods, fostering a oom-robust ecosystem just as we have a burgeoning no_std ecosystem.

**Comment by Manishearth** [comment]
So, I'm probably missing something, but is there a similar RFC/tracking issue/etc for `Box::try_new`? Is there some place I can track progress for this?

**Comment by SimonSapin** [comment]
> #77118 has been merged. Can stabilization of this proceed?

In the June 2019 to October 2019 part of this thread, the part of custom allocators that was deemed affecting the design of `try_resreve` is error types.

That is, if error values for allocation failures (potentially) carry meaningful information then the error type of `try_reserve` should forward that information to its own callers. In particular if the allocator trait ends up with an associated type for its own errors then the error type of `try_reserve` should be *generic*.

By this logic (which my 2019 proposal did not adhere to), `try_reserve` should only be stabilized after the allocation trait is stabilized. #77118 is only one of the steps required for that.

---

> is there a similar RFC/tracking issue/etc for `Box::try_new`?

By the letter of RFC 2116 I believe there is not. That RFC only proposed `try_reserve` for collection types in order to keep its scope small and not get into the design of `try_push`, `try_extend`, etc.

In the spirit of that RFC however, `try_new` is the minimal viable API for fallible allocation in non-collection containers in the same way that `try_reserve` is for containers. So I’d personally be fine with including it here.

**Comment by Manishearth** [comment]
> In the spirit of that RFC however, `try_new` is the minimal viable API for fallible allocation in non-collection containers in the same way that `try_reserve` is for containers. So I’d personally be fine with including it here.

I'm talking to some folks doing Linux kernel work in Rust who would like to see this. Should I just open a PR for it, feature gate it with `try_new` (or `try_reserve`?) for the same tracking issue?

**Comment by Ericson2314** [comment]
@Manishearth maybe https://github.com/rust-lang/wg-allocators/issues/7 might help. You could at least see what unstable annotations those PRs used.

**Comment by baumanj** [comment]
> I'm talking to some folks doing Linux kernel work in Rust who would like to see this. Should I just open a PR for it, feature gate it with `try_new` (or `try_reserve`?) for the same tracking issue?

In the meantime, you may be interested in (or already know about) https://crates.io/crates/fallible_collections. We've been using (and contributing to) that crate for awhile now to support the rust MP4 parser that Gecko uses. I've found it quite nice to work with.

**Comment by Manishearth** [comment]
Opened https://github.com/rust-lang/rust/pull/80310

**Comment by Stargateur** [comment]
BTW, doesn't this could also add:

- `try_with_capacity()`
- `try_push()` 
- `try_append()`
- `try_insert()`
- `try_split_off()`
- `try_resize_with()`
- `try_resize()`
- `try_extend_from_slice()`

I guess I miss some, this will add a lot of method on `Vec` but I think there is no reason they should not follow. I wonder if we could put all of this in a trait just for `Vec`.

**Comment by Ericson2314** [comment]
@Stargateur it does indeed work, and is nice and ergonomic with `?`.

**Comment by the8472** [comment]
> That is, if error values for allocation failures (potentially) carry meaningful information then the error type of `try_reserve` should forward that information to its own callers. In particular if the allocator trait ends up with an associated type for its own errors then the error type of `try_reserve` should be _generic_.
> 
> By this logic (which my 2019 proposal did not adhere to), `try_reserve` should only be stabilized after the allocation trait is stabilized. #77118 is only one of the steps required for that.

Would using an opaque type be possible, i.e. `Result<_, impl Error>`? Could that be replaced with `<A as Allocator>::Error` or `AllocError` later? That would decouple the allocator API question from the `try_` effort.

**Comment by kornelski** [comment]
Has anyone proposed a use-case for the `TryAllocError`?

The discussion in this thread so far is only focused on what could be done with the error, but I don't see justification why it's being done and who is it for. It feels like a solution in search of a problem to solve. 

In what situations it's necessary to handle the error differently depending on the details like `Layout`?

Why would any application care whether it's a `CapacityOverflow` or actual OOM? Isn't this distinction just formalizing a quirk/bug in the current Rust implementation?

**I suggest using zero-sized type** `TryAllocError`, and not giving *any* details at all. 

* Applications that embrace fallible allocations will use them a lot, possibly in hot loops with `try_push`, and a complex error type would add overhead, e.g. look how much bloat `&mut Vec as io::Write` adds for just moving the `io::Error` around.

* `malloc` just returns `NULL` and AFAIK nobody ever complained about the lack of detail there. There's no `errno` equivalent for it.

* C++ `bad_alloc` exception also has no useful info ("bad allocation" string is all you get).

* When I've used the `fallible_collections` crate and I didn't need any details from the error. In some cases I wanted to report something more informational than "out of memory", but for this I actually needed higher-level application-domain info like "Not enough memory for a {}x{} pixels large image", and `Layout` is not useful for this — it's too low-level.

* In multithreaded applications any information from the allocator is inherently racy and subject to TOCTOU errors. If you imagine some use-case like checking the error details to decide whether to retry with a smaller allocation - that won't work. By the time you check the error the data may already be out of date. If you want to know whether another allocation would have succeeded, then the *only* way is to actually perform the allocation.

So I think for simplicity and performance zero-sized `TryAllocError` is the best.

**Comment by Ericson2314** [comment]
Could we do a ZST `impl trait`? That would be even more opaque.

**Comment by Stargateur** [comment]
> Why would any application care whether it's a `CapacityOverflow` or actual OOM? Isn't this distinction just formalizing a quirk/inconsistency in the current Rust implementation?

I disagree, capasity overflow is more a limitation of Rust than a limitation of the hardware and so it's important difference with pure OOM. The problem with most of your following point is that your didn't take in account that C and C++ don't have this limitation and so don't need to make a difference and so your comparative like:

> malloc just returns NULL and AFAIK nobody ever complained about the lack of detail there. There's no errno equivalent for it.

are pointless.

**Comment by comex** [comment]
> C and C++ don't have this limitation

If you're referring to allocations having to fit in `isize` rather than `usize`, C and C++ do have that limitation.  While the C standard doesn't explicitly ban allocations greater than half the address space, it defines pointer subtraction as returning the signed type `ptrdiff_t`.  Since pointer subtraction is only defined for pointers within the same allocation, this leads to a lot of problems, [including miscompilations due to compiler assumptions](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67999), if and only if a single allocation can be larger than `PTRDIFF_MAX` bytes.  For this reason, some implementations of libc have long capped allocations at that size, and [glibc added such a cap in 2019](https://lwn.net/Articles/795127/).  

But I believe `CapacityOverflow` would also cover the case where multiplying the capacity by the element size simply overflows `usize` entirely.  That could be seen as more of a hardware limitation.

C++ actually does distinguish capacity overflow (of either kind) from other errors, by throwing a different exception.  C++ standard containers have a [`max_size()` method](https://en.cppreference.com/w/cpp/container/vector/max_size), returning the maximum possible size that can be allocated.  If you call [`reserve`](https://en.cppreference.com/w/cpp/container/vector/reserve) with a value higher than `max_size()`, you get `std::length_error`, as opposed to `std::bad_alloc` for other allocation failures.

On the other hand, C doesn't distinguish them.  While C doesn't have standard containers, it does have the `calloc` function, which accepts a count and an element size, and multiplies them together to determine how many bytes to allocate.  This multiplication is overflow-checked, but overflow results in `calloc` returning `NULL` with errno set to `ENOMEM`, same as other allocation failures.

**Comment by kornelski** [comment]
but what would an application *do* with that info?

```rust
match vec.try_reserve(size) {
   Err(CapacityOverflow) => {…} // A
   Err(AllocErr(_)) => {…} // B
   _ => {},
}
```

what app needs A case to be different from B?

I'm struggling to come up with anything, because in both cases the end result is the same — you just can't allocate the size you wanted. And in both cases you can't assume that you will be able to allocate some other size until you try to.

And even if an application wanted to allocate as close to half of the address space as it can get, then it doesn't need the error, it just needs to compute the correct size ahead of time from `size_of` and `isize::MAX`. And if we're dealing with complex containers that may have arbitrary capacity limitations, then they probably should expose something like `Collection::MAX_LEN` to allow callers to choose acceptable capacity before getting the error, instead of probing it through repeated failures.

**Comment by Stargateur** [comment]
> While the C standard doesn't explicitly ban allocations greater than half the address space

I was talking about C standard, I don't care of C implementation specificity. My point remain. I agree with the rest of your message.

> but what would an application _do_ with that info?

Do you really handle type of error in your application, in almost 99% of application the way to handle error is to return it in cascade. The real value is to known what is the error for the real user and maybe fix the code. Or change application parameter to avoid allocate too much on this specific example. If a user see OOM, the obvious fix is too add RAM but adding RAM when you have hit a software limitation will not help at all. I agree this will not happen anytime soon for 64 bit. But that would be nice to have it for 32 bit system and other thing like embedded or boot code.

> I'm struggling to come up with anything, because in both cases the end result is the same — you just can't allocate the size you wanted. And in both cases you can't assume that you will be able to allocate some other size until you try to.

With this logic why bother report any error anyway what the application will do ? let's just make all API return "An error have happen".

> Even in case of a hypothetical container with non-trivial maximum capacity, this error is also unnecessary. Such collection should expose something like `Collection::MAX_LEN` to allow callers to choose an acceptable capacity before getting the error, instead of probing it through failures.

Like std ? I fail to remember such feature exist for vector. So I must call an function myself to verify I do not send wrong value before I call the real function I want to use ? This is not user friendly.

**Comment by comex** [comment]
> but what would an application _do_ with that info?

Yeah, I personally agree that this information is not very useful.

**Comment by kornelski** [comment]
I've got an idea for an express path to stabilization of `try_reserve` — hide it in the `io::Write` implementation! `vec.write_all(s)` is equivalent to `vec.try_extend_from_slice(s)`.

https://github.com/rust-lang/rust/pull/84612

**Comment by nbdd0121** [comment]
In addition to @kornelski's ZST suggestion, I personally think we shouldn't even have separate `TryReserveError` and `AllocError`.

**Comment by kornelski** [comment]
Now TryReserveError is an opaque type. 
Besides the error type, were there any other blockers to stabilization?

**Comment by TimDiekmann** [comment]
As far as I know, that was the only blocker.

**Comment by kornelski** [comment]
# Request for stabilization

I'd like to propose to stabilize this in 1.56.

### Brief summary

`try_reserve()` and `try_reserve_exact()` methods on `Vec`, `VecDeque`, `HashMap`, `HashSet`, and `String`. These methods enable best-effort graceful handling of out of memory errors, especially large allocations on operating systems without virtual memory overcommit, or when using custom allocators that self-impose memory usage limits (like [cap](https://crates.io/crates/cap)).

```rust
let mut vec = Vec::new();
vec.try_reserve(lots)?;
/// now pushing lots of bytes won't abort
```

### Testing

Unit [tests are in liballoc](https://github.com/rust-lang/rust/blob/4498e300e41f47c75abe4e49ec91ae949aaeea5f/library/alloc/tests/vec.rs#L1459). Fallible allocation in general has been prototyped via [fallible_collections](https://crates.io/crates/fallible_collections) crate, e.g. shipped in the mp4 parser used in Firefox.

### Questions intentionally left unresolved

The `TryReserveError` type is left as an opaque struct in order to leave options open for possible future extensions to custom allocators, or it may be later reduced to a unit struct if it turns out that there's no user demand for exposing further details.

**Comment by nbdd0121** [comment]
Do we have a way to reserve the possibility of merging `AllocError` and `TryReserveError` into the same type?

**Comment by kornelski** [comment]
It could wrap it, or:

```rust
pub type TryReserveError = AllocError;
```

**Comment by nbdd0121** [comment]
If we ended up wanting `TryReserveError` to not distinguish between CapacityOverflow and AllocError, then having `TryReserveError` and `AllocError` as separate type would be unergonomic IMO. On the other hand a type alias would break trait impls.

**Comment by the8472** [comment]
AllocError is unstable, so we could merge them before it is stabilized.

**Comment by the8472** [comment]
> Request for stabilization

In a libs team meeting there was the discussion of having a separate `Vec` type that contains all the fallible methods and cheap conversion methods between the fallible and infallible Vec. If that were implemented then `try_reserve` on what would then be the infallible Vec would be redundant.

So the question would be if that kind of overlap is acceptable to get it stabilized now or whether the separate Vec type should be pursued instead.

**Comment by kornelski** [comment]
@the8472 I've tried to use the [`TryVec` type](https://docs.rs/fallible_collections/0.4.3/fallible_collections/vec/struct.TryVec.html) in a large server project, and I found it [*completely unworkable*](https://internals.rust-lang.org/t/feedback-from-adoption-of-fallible-allocations/14502). That approach was a dead-end, and I ended up deleting that branch and staring over with only regular types and `try_reserve`. 

It was both too restrictive and too lax. Too restrictive: incompatibility with `Vec` made small type changes cascade throughout the entire codebase, and clash with dependencies. It was annoying and needless churn, because not every use of `Vec` needs to grow it. Storing/returning non-growable buffers as a `Vec` is important, because conversion to a boxed slice often re-allocates.

And despite highly disruptive viral nature of the type, it wasn't enough: It doesn't stop code from using aborting `Vec` for temporaries in functions and in private fields. It doesn't stop use of `vec!`, `Vec::new` or `collect()` if such use doesn't affect other code through public fields and return types. I was really deeply disappointed that even after I've upturned my whole codebase to change a fundamental type, it was still aborting like before.

So I think having *only* a separate type without an easy-to-adopt `try_reserve` would be highly problematic for usability. A different fallible-only type could be pursued regardless, but such approach is not a replacement for what `try_reserve` does, so I urge you to stabilize `try_reserve`.

**Comment by TimDiekmann** [comment]
The current idea in the allocator WG is to use the `no_global_oom_handling` feature: https://github.com/rust-lang/rust/pull/84266. When this feature is set, all fallible allocations are disabled and only the `try_` variants will be available. This doesn't introduce a new type and is fully compatible.

**Comment by kornelski** [comment]
`no_global_oom_handling` is basically only for the Linux kernel and embedded cases. It doesn't satisfy server and application use-cases from the RFC, because it's incompatible with all std-dependent crates on crates.io. It's not an alternative to `try_reserve`.

**Comment by TimDiekmann** [comment]
No, it's not an alternative, but it plays together very well. I expect the usual use case to use `try_reserve` in critical section if needed but also use `reserve` (or any method calling it) in the same program. As you pointed out, splitting `Vec` into two types will make this very hard. Splitting `Vec` was also discussed in the allocator WG at Zulip (Can't find the link currently) and ended up with different conversion methods. It was discussed because some libraries (i.e. the Linux kernel) must not use `reserve` and `TryVec` would prohibit the usage. For such crates, `no_global_oom_handling` is great without adding extra types to std.

For me, it totally makes sense to add `try_reserve` as an option.

**Comment by Stargateur** [comment]
@kornelski I agree with I don't really care about detail information but I **strongly** think I want to be able to differentiate a `OOM` error from a `too big for isize` error.

**Comment by Stargateur** [comment]
@TimDiekmann I think this should be added as a form of a trait to avoid too many `try_` method or maybe a intermediate type that only have failure handling method. I still hate the original choice to not have return an error for `OOM` see my 2017 SO question https://stackoverflow.com/q/44923306/7076153.

```rust
impl Vec {
    fn find_a_name(&mut self) -> TryVec;
}

impl TryVec {
    fn push(&mut self) -> Result;
    // basically all vector method
```

**Comment by the8472** [comment]
@kornelski 

> @the8472 I've tried to use the [`TryVec` type](https://docs.rs/fallible_collections/0.4.3/fallible_collections/vec/struct.TryVec.html) in a large server project, and I found it [_completely unworkable_](https://internals.rust-lang.org/t/feedback-from-adoption-of-fallible-allocations/14502). That approach was a dead-end, and I ended up deleting that branch and staring over with only regular types and `try_reserve`.
> 
> It was both too restrictive and too lax. Too restrictive: incompatibility with `Vec` made small type changes cascade throughout the entire codebase, and clash with dependencies. It was annoying and needless churn, because not every use of `Vec` needs to grow it. Storing/returning non-growable buffers as a `Vec` is important, because conversion to a boxed slice often re-allocates.

I think that is mostly a limitation of `TryVec` because it doesn't live in the standard library. If they are located in the same crate then you can define interconversion-methods that turn a `&mut Vec` into a `&mut FallibleVec` and vice versa which you can pass to any API expecting those. And of course the owned types can also be turned into each other. And that of course means they can also be turned into `Box<[T]>` if needed.

> And despite highly disruptive viral nature of the type, it wasn't enough: It doesn't stop code from using aborting `Vec` for temporaries in functions and in private fields. It doesn't stop use of `vec!`, `Vec::new` or `collect()` if such use doesn't affect other code through public fields and return types. I was really deeply disappointed that even after I've upturned my whole codebase to change a fundamental type, it was still aborting like before.

That's what `no_global_oom_handling` would achieve, it would basically forbid using the infallible version of Vec.

> So I think having _only_ a separate type without an easy-to-adopt `try_reserve` would be highly problematic for usability. A different fallible-only type could be pursued regardless, but such approach is not a replacement for what `try_reserve` does, so I urge you to stabilize `try_reserve`.

I think this isn't true if you have free conversion between the types.

----

@TimDiekmann 

> The current idea in the allocator WG is to use the `no_global_oom_handling` feature: #84266. When this feature is set, all fallible allocations are disabled and only the `try_` variants will be available. This doesn't introduce a new type and is fully compatible.

Major issues with that design are a) you have to slap the feature gate on everything. It's annoying for maintenance. Having separate types would make this much easier since you could gate entire modules.

Additionally it makes naming easier because you wouldn't have to prefix everything with `try_`, you could just have two versions of `reserve`, `push` etc. on the different types, one that returns a `Result` and the one that doesn't. I.e. it simplifies naming.


Anyway, this is just my what I understood from the libs team meeting. Maybe those were just weakly held ideas. :man_shrugging:

**Comment by ojeda** [comment]
> I think that is mostly a limitation of `TryVec` because it doesn't live in the standard library. If they are located in the same crate then you can define interconversion-methods that turn a `&mut Vec` into a `&mut FallibleVec` and vice versa which you can pass to any API expecting those. And of course the owned types can also be turned into each other. And that of course means they can also be turned into `Box<[T]>` if needed.

But what is gained by splitting the types? In particular, if there are safe, cheap interconversion methods, then I assume the types will not guarantee their allocations have been / will be fallible or infallible. Thus the split would be mostly used as a way to namespace methods, right? And, if those names end up being the same for both as you propose, as a fun way to perform method overloading.

> Major issues with that design are a) you have to slap the feature gate on everything. It's annoying for maintenance. Having separate types would make this much easier since you could gate entire modules.

While that is important to consider, in the end it is an implementation detail.

> Additionally it makes naming easier because you wouldn't have to prefix everything with `try_`, you could just have two versions of `reserve`, `push` etc. on the different types, one that returns a `Result` and the one that doesn't. I.e. it simplifies naming.

Yes, names would be simpler if a project only uses a single kind of vector everywhere, but for any project that happens to mix them, it could end up being more confusing -- similar to supporting overloading.

**Comment by the8472** [comment]
> But what is gained by splitting the types? In particular, if there are safe, cheap interconversion methods, then I assume the types will not guarantee their allocations have been / will be fallible or infallible. 

You mean enforcing that *other* libraries don't do infallible allocation? Yeah, that's not the goal here. The goal is to split the humongous API surface of `Vec`. Having a separate type and conversion methods makes it easy to isolate the infallible parts locally. Make sure that all your oom-panic-free modules don't contain `Vec<>` or `to_infallible()` and that should be it, the fallible convenience feature could use them in a separate module that's exempt from that.

On top of that `no_global_oom_handling` would of course still disable the infallible versions.

**Comment by Lokathor** [comment]
If that's not a goal I wouldn't bother with a split at all. Having all allocating methods also have a `try_` variant doesn't meaningfully increase the API surface at all. You just explain *once* what `try_` means and the user is all set.

Compare with integers: Adding a method to all signed integer types is technically affecting 6 different types, but users actually still just think of it as "one" new operation you can do.

**Comment by kornelski** [comment]
I can't imagine using a global kill-switch `no_global_oom_handling`. My server/application projects have (transitively) *hundreds* of dependencies. I will not be able convince maintainers of every single crate to switch to fallible allocations *everywhere*. Even if a single dependency, even in one place, uses some infallible feature it would stop my whole application from using `no_global_oom_handling`. And for a dependency-of-dependency-of-dependency it may make perfect sense to e.g. use `format!()`, and they wouldn't care that I can't have them use `format!()` because my project elsewhere does something else that needs to handle OOM.

Besides, I don't even want fallible allocations everywhere in my projects. For example, my servers have substantial startup code that doesn't need to handle OOM.

**Comment by kornelski** [comment]
@Stargateur @the8472 The traits and methods you're discussing have already been considered two years ago when [the fallible allocation RFC has been accepted](https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md). The RFC opt *not* to provide a separate fallible Vec type, nor add try_ methods for everything, because `try_reserve` is the minimal building block that can be used to create all of these in user code.

**Comment by Stargateur** [comment]
@kornelski and still there seem to be not a consensus "In a libs team meeting there was the discussion of having a separate Vec type that contains all the fallible methods", so I think it's still open in fact, or let just stop for now to wait and just do what we agree on in the RFC ? why are we waiting so ? Why are we talking ?

**Comment by the8472** [comment]
I asked on zulip, [the response](https://rust-lang.zulipchat.com/#narrow/stream/219381-t-libs/topic/alloc.20non-global-oom.20support/near/249408555) was:

> Several people in that discussion very much *didn't* want to see separate types, because those types tend to "infect" everything in an API, and it's a pain to support both.

**Comment by dtolnay** [comment]
What is the motivation for `impl From<LayoutError> for TryReserveError`? Could someone share an example of what a realistic use case would look like with and without this impl to show what difference it makes?

It doesn't seem straightforward that a `LayoutError` due to e.g. non-power-of-2 alignment would turn into a "capacity overflow" after the conversion.

**Comment by kornelski** [comment]
Is there a use-case for *creating* instances of `TryReserveError`? 

One I can think of is someone creating a custom data structure, and wanting to mimic std API. But that's not a very strong case. 3rd party crates can use `io::ErrorKind::OutOfMemory` or just their own error type.

If there's no need to be able to create `TryReserveError` instances in user code, then it would be fine to remove the `From` impl.

**Comment by SimonSapin** [comment]
I believe this `From` impl is used by the standard library’s implementation of `try_reserve`. Yes it could be removed and replaced by a manual conversion.

**Comment by orlp** [comment]
It's super minor but `TryReserveErrorKind` is missing periods at the end of sentences in the documentation.

**Comment by 8573** [comment]
> It's super minor but `TryReserveErrorKind` is missing periods at the end of sentences in the documentation.

The only such sentence I see there is "The memory allocator returned an error"; the others are just sentence fragments, which may not deserve periods <small>(for example, where sentence fragments may be used in Wikipedia, their style guide (probably the result of compromise between many strong opinions) [says](https://en.wikipedia.org/wiki/MOS:PERIOD) these "should in most cases not end with a period")</small>.

**Comment by orlp** [comment]
> the others are just sentence fragments, which may not deserve periods

https://doc.rust-lang.org/1.0.0/style/style/comments.html#sentence-structure

> All doc comments, including the summary line, should begin with a capital letter and end with a period, question mark, or exclamation point. Prefer full sentences to fragments.

Sentence fragments should end with a period in Rust, even if full sentences are preferred..

**Comment by safinaskar** [comment]
Important news! `HashMap` got method `try_insert` ( https://github.com/rust-lang/rust/issues/82766 ), where `try_` means something completely different to `Box::try_new`. Please, prevent somehow this `try_insert` from becoming stable

**Comment by Stargateur** [comment]
I think [`insert_if_vacant`](https://github.com/rust-lang/rfcs/issues/3092#issuecomment-790788137) is way better, `try` is quite confusing with allocation error on collection. Also, try_ is too much generic on this case of specific "only if vacant".

The downside of this is it's rather long. Since even `VacantEntry` use [`insert()`](https://doc.rust-lang.org/std/collections/hash_map/struct.VacantEntry.html#method.insert) I think we should keep `insert()` key word.

My proposition is `insert_vacant()`. (aka insert if vacant entry).

**Comment by SimonSapin** [comment]
Please file an issue dedicated to the naming of `try_insert` since this thread is already covering other topics.

**Comment by Toru3** [comment]
The comment of example code is seems wrong.
If size of `data` is about 60% of system memory, `try_reserve` might success because OS doesn't allocate physical memory immediately.
Writing to allocated memory causes page fault and OS allocates memory.
Sum of size of old data and new data is about 120% of system memory.
So, OOM might occur.

```
$ rg -C3 "can't OOM"
library/std/src/ffi/os_str.rs
311-    ///     // Pre-reserve the memory, exiting if we can't
312-    ///     s.try_reserve(OsStr::new(data).len())?;
313-    ///
314:    ///     // Now we know this can't OOM in the middle of our complex work
315-    ///     s.push(data);
316-    ///
317-    ///     Ok(s)
--
382-    ///     // Pre-reserve the memory, exiting if we can't
383-    ///     s.try_reserve_exact(OsStr::new(data).len())?;
384-    ///
385:    ///     // Now we know this can't OOM in the middle of our complex work
386-    ///     s.push(data);
387-    ///
388-    ///     Ok(s)

library/alloc/src/string.rs
1098-    ///     // Pre-reserve the memory, exiting if we can't
1099-    ///     output.try_reserve(data.len())?;
1100-    ///
1101:    ///     // Now we know this can't OOM in the middle of our complex work
1102-    ///     output.push_str(data);
1103-    ///
1104-    ///     Ok(output)
--
1139-    ///     // Pre-reserve the memory, exiting if we can't
1140-    ///     output.try_reserve_exact(data.len())?;
1141-    ///
1142:    ///     // Now we know this can't OOM in the middle of our complex work
1143-    ///     output.push_str(data);
1144-    ///
1145-    ///     Ok(output)

library/alloc/src/collections/binary_heap.rs
994-    ///     // Pre-reserve the memory, exiting if we can't
995-    ///     heap.try_reserve_exact(data.len())?;
996-    ///
997:    ///     // Now we know this can't OOM in the middle of our complex work
998-    ///     heap.extend(data.iter());
999-    ///
1000-    ///     Ok(heap.pop())
--
1029-    ///     // Pre-reserve the memory, exiting if we can't
1030-    ///     heap.try_reserve(data.len())?;
1031-    ///
1032:    ///     // Now we know this can't OOM in the middle of our complex work
1033-    ///     heap.extend(data.iter());
1034-    ///
1035-    ///     Ok(heap.pop())

library/alloc/src/collections/vec_deque/mod.rs
776-    ///     // Pre-reserve the memory, exiting if we can't
777-    ///     output.try_reserve_exact(data.len())?;
778-    ///
779:    ///     // Now we know this can't OOM(Out-Of-Memory) in the middle of our complex work
780-    ///     output.extend(data.iter().map(|&val| {
781-    ///         val * 2 + 5 // very complicated
782-    ///     }));
--
813-    ///     // Pre-reserve the memory, exiting if we can't
814-    ///     output.try_reserve(data.len())?;
815-    ///
816:    ///     // Now we know this can't OOM in the middle of our complex work
817-    ///     output.extend(data.iter().map(|&val| {
818-    ///         val * 2 + 5 // very complicated
819-    ///     }));

library/alloc/src/vec/mod.rs
893-    ///     // Pre-reserve the memory, exiting if we can't
894-    ///     output.try_reserve(data.len())?;
895-    ///
896:    ///     // Now we know this can't OOM in the middle of our complex work
897-    ///     output.extend(data.iter().map(|&val| {
898-    ///         val * 2 + 5 // very complicated
899-    ///     }));
--
936-    ///     // Pre-reserve the memory, exiting if we can't
937-    ///     output.try_reserve_exact(data.len())?;
938-    ///
939:    ///     // Now we know this can't OOM in the middle of our complex work
940-    ///     output.extend(data.iter().map(|&val| {
941-    ///         val * 2 + 5 // very complicated
942-    ///     }));
```

**Comment by kornelski** [comment]
@Toru3 Overcommit is a known limitation, and it's been considered in [the RFC](https://github.com/rust-lang/rfcs/blob/master/text/2116-alloc-me-maybe.md#additional-background-overcommit-and-killers). You can mitigate the problem on the OS level either by disabling overcommit, or running code in containers with limited amount of memory. Within Rust programs you can reduce chance of hitting the problem by using the [`cap` allocator](//lib.rs/cap) with a limit below the amount of physical memory available.

**Comment by studyingegret** [comment]
Just to check my understanding, are `TryReserveError::kind` and `TryReserveErrorKind` the only blockers now?

**Comment by abrahem79** [comment]
```

impl /* each of String, Vec<T>, VecDeque<T> */ {
    pub fn try_reserve(&mut self, additional: usize) -> Result<(), TryReserveError> {…}
    pub fn try_reserve_exact(&mut self, additional: usize) -> Result<(), TryReserveError> {…}
}

impl /* each of HashMap<K, V> and HashSet<T> */ {
    pub fn try_reserve(&mut self, additional: usize) -> Result<(), TryReserveError> {…}
}

/// The error type for `try_reserve` methods.
#[derive(Clone, PartialEq, Eq, Debug)]
pub enum TryReserveError { // in std::collections
    /// Error due to the computed capacity exceeding the collection's maximum
    /// (usually `isize::MAX` bytes).
    CapacityOverflow,

    /// The memory allocator returned an error
    AllocError {
        /// The layout of allocation request that failed
        layout: Layout,

        #[doc(hidden)]
        #[unstable(feature = "container_error_extra", issue = "0", reason = "\
            Enable exposing the allocator’s custom error value \
            if an associated type is added in the future: \
            https://github.com/rust-lang/wg-allocators/issues/23")]
        non_exhaustive: (),
    },
}

impl From<LayoutErr> for TryReserveError {
    fn from(_: LayoutErr) -> Self {
        TryReserveError::CapacityOverflow
    }
}

```
